File: src/main/scala/nebflow/gateway/SttService.scala
package nebflow.gateway

import cats.effect.IO
import io.circe.{Json, JsonObject}
import io.circe.parser.parse
import io.circe.syntax.*
import nebflow.core.{NebflowLogger, PathUtil}
import nebflow.shared.Defaults

import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Path}
import java.util.Base64

object SttService:
  private val logger = NebflowLogger.forName("nebflow.stt")

  /** STT 配置文件路径。def（非 val）——PathUtil.dataRoot 可被测试 setDataRoot
    * 重定向，val 会在类加载时冻结（#295）。 */
  def configPath: os.Path = PathUtil.dataRoot / "stt-config.json"

  def create(): IO[Option[SttService]] =
    IO.blocking {
      if !os.exists(configPath) then None
      else
        parse(os.read(configPath)).toOption.flatMap { json =>
          val hc = json.hcursor
          val apiKeyOpt = hc.downField("apiKey").as[String].toOption.filter(_.nonEmpty)
          apiKeyOpt.flatMap { apiKey =>
            val model = hc.downField("model").as[String].getOrElse(Defaults.SttDefaultModel)
            val endpoint = hc
              .downField("endpoint")
              .as[String]
              .getOrElse(Defaults.SttDefaultEndpoint)
            Some(new SttService(apiKey, model, endpoint))
          }
        }
    }.flatMap {
      case Some(svc) => logger.info("STT service initialized").as(Some(svc))
      case None => IO.pure(None)
    }
  end create

  /** #295 A2：字段级三态。None=字段省略（保留旧值）；Some(None)=显式空串
    * （清除该字段）；Some(Some(v))=设值（已 trim、非空）。 */
  type SttField = Option[Option[String]]

  final case class SttConfigPatch(endpoint: SttField, apiKey: SttField, model: SttField):
    /** 全部字段显式清除——等价于清空整个配置。 */
    def clearsEverything: Boolean =
      List(endpoint, apiKey, model).forall(_.contains(None))

  /**
   * setSttConfig WS 命令的校验/解析入口（#295 → A2 部分更新语义，纯函数）。
   *
   * 语义（2026-08-20 A2 对齐，前端契约：空 key 输入框省略字段）：
   * - 字段省略 → 保留旧值（merge 侧处理，本次不触碰）
   * - 字段显式空串/纯空白 → 清除该字段
   * - 字段设值 → 校验后写入（endpoint 须 http(s)://；model/apiKey 非空即可）
   *
   * 旧 validateConfig 把「省略」与「显式空」折叠成同一路径且整文件替换，
   * 导致二次保存（前端 apiKey 恒空、省略该字段）时旧 key 被覆盖丢失。
   */
  def parsePatch(payload: Json): Either[String, SttConfigPatch] =
    def field(name: String): Either[String, SttField] =
      payload.hcursor.downField(name).as[Option[String]] match
        case Left(_) => Left(s"STT field '$name' must be a string if present")
        case Right(None) => Right(None) // field absent from JSON — keep old value
        case Right(Some(raw)) =>
          val v = raw.trim
          if v.isEmpty then Right(Some(None)) // explicit empty — clear this field
          else Right(Some(Some(v)))
    for
      endpoint <- field("endpoint")
      apiKey <- field("apiKey")
      model <- field("model")
      _ <- endpoint.flatten match
        case Some(e) if !e.startsWith("http://") && !e.startsWith("https://") =>
          Left(s"STT endpoint must start with http:// or https:// (got: ${e.take(60)})")
        case _ => Right(())
    yield SttConfigPatch(endpoint, apiKey, model)

  /**
   * patch 叠加旧配置（#295 A2 纯函数）。省略=保留旧值；显式空=删字段；
   * 设值=覆写。返回 None=合并后无任何字段（未配置态——调用方删文件，
   * 服务回退浏览器 Web Speech）。部分字段存在但缺 apiKey 时 create()
   * 返回 None（无代理服务），字段保留在盘上待 key 回填。
   */
  def mergeConfig(old: Option[Json], patch: SttConfigPatch): Option[Json] =
    def oldField(name: String): Option[String] =
      old.flatMap(_.hcursor.downField(name).as[String].toOption).filter(_.nonEmpty)
    def apply(name: String, f: SttField): Option[(String, Json)] =
      f match
        case None         => oldField(name).map(v => name -> v.asJson) // omitted — keep
        case Some(None)   => None // explicit clear
        case Some(Some(v)) => Some(name -> v.asJson)
    val fields = List(
      apply("endpoint", patch.endpoint),
      apply("apiKey", patch.apiKey),
      apply("model", patch.model)
    ).flatten
    if fields.isEmpty then None
    else Some(Json.fromJsonObject(JsonObject.fromIterable(fields)))

  /**
   * serverConfig 广播的 stt 节（#295）：{sttConfigured, endpoint?, model?}——
   * **apiKey 永不外露**（key 不落地前端铁律）。未配置时仅 sttConfigured:false。
   */
  def serverConfigNode(svc: Option[SttService]): Json =
    svc match
      case Some(s) =>
        Json.obj(
          "sttConfigured" -> true.asJson,
          "endpoint" -> s.endpoint.asJson,
          "model" -> s.model.asJson
        )
      case None =>
        Json.obj("sttConfigured" -> false.asJson)
end SttService

/**
  * STT 语音识别服务。按 endpoint 后缀自动识别两种协议（2026-08-21 MiMo 支持）：
  *
  * 1. **chat/completions 协议**——endpoint 以 `/chat/completions` 结尾。MiMo 的
  *    「OpenAI API Compatibility」是 chat 协议而非 transcriptions（无
  *    /v1/audio/transcriptions，全部 404）。实测定稿：input_audio data-URL
  *    (WAV base64) + asr_options.language（仅 auto/zh/en），转写文本在
  *    choices[0].message.content。
  * 2. **multipart transcriptions 协议**（原有路径）——OpenAI Whisper 风格服务。
  *
  * 不加新配置字段，按 endpoint 后缀分流。配置文件: ~/.nebflow/stt-config.json
  * {
  *   "apiKey": "your-key",
  *   "model": "mimo-v2.5-asr",
  *   "endpoint": "https://api.xiaomimimo.com/v1/chat/completions"
  * }
  */
class SttService private[gateway] (
  apiKey: String,
  /** 转录模型名（serverConfig 广播可见；apiKey 永不外露）。 */
  val model: String,
  /** 转录端点（serverConfig 广播可见；apiKey 永不外露）。 */
  val endpoint: String
):
  private val logger = NebflowLogger.forName("nebflow.stt")

  private val client = HttpClient
    .newBuilder()
    .connectTimeout(java.time.Duration.ofSeconds(10))
    .build()

  /**
   * 转录 WAV 音频数据，返回识别文本。
   *
   * @param wavBytes WAV 格式的音频字节数组（前端固定 16kHz mono WAV）
   * @param language 语言代码（前端传 BCP-47 如 zh-CN/en-US；两协议各自映射）
   * @return Right(识别文本) 或 Left(错误消息，回传前端)
   */
  def transcribe(wavBytes: Array[Byte], language: Option[String]): IO[Either[String, String]] =
    if wavBytes.isEmpty then IO.pure(Left("Empty audio data"))
    else if isChatCompletionsEndpoint then transcribeViaChat(wavBytes, language)
    else transcribeViaMultipart(wavBytes, language)
  end transcribe

  /** chat/completions 协议判定：仅按 endpoint 后缀（trim 后），无新配置字段。 */
  private def isChatCompletionsEndpoint: Boolean =
    endpoint.trim.endsWith("/chat/completions")

  /** chat 协议语言映射：MiMo asr_options.language 仅接受 auto/zh/en。前端传
    * BCP-47（zh-CN/en-US/...）——startsWith 前缀映射，其余（含 None）→ auto。 */
  private def mapLanguage(language: Option[String]): String =
    language.map(_.trim.toLowerCase) match
      case Some(l) if l.startsWith("zh") => "zh"
      case Some(l) if l.startsWith("en") => "en"
      case _                             => "auto"

  /** MiMo 风格 chat/completions ASR（2026-08-21 实测定稿，0.8s 中文转写验证）。 */
  private def transcribeViaChat(
      wavBytes: Array[Byte],
      language: Option[String]
  ): IO[Either[String, String]] =
    val audioB64 = Base64.getEncoder.encodeToString(wavBytes)
    val body = Json.obj(
      "model" -> model.asJson,
      "messages" -> Json.arr(
        Json.obj(
          "role" -> "user".asJson,
          "content" -> Json.arr(
            Json.obj(
              "type" -> "input_audio".asJson,
              "input_audio" -> Json.obj(
                "data" -> s"data:audio/wav;base64,$audioB64".asJson,
                "format" -> "wav".asJson
              )
            )
          )
        )
      ),
      "asr_options" -> Json.obj("language" -> mapLanguage(language).asJson)
    )

    val request = HttpRequest
      .newBuilder()
      .uri(URI.create(endpoint.trim))
      .header("Authorization", s"Bearer $apiKey")
      .header("Content-Type", "application/json")
      .POST(HttpRequest.BodyPublishers.ofString(body.noSpaces, StandardCharsets.UTF_8))
      .timeout(java.time.Duration.ofSeconds(30))
      .build()

    IO.blocking {
      val response = client.send(request, HttpResponse.BodyHandlers.ofString())
      if response.statusCode() != 200 then
        // 语句位在 IO.blocking 内——必须 *Sync（裸 logger.warn 返回 IO[Unit] 被丢弃=死日志）
        logger.warnSync(s"STT API error ${response.statusCode()}: ${response.body().take(300)}")
        Left(s"STT API error: ${response.statusCode()}")
      else
        parse(response.body()).toOption match
          case Some(json) =>
            json.hcursor
              .downField("choices").downArray
              .downField("message").downField("content").as[String] match
              case Right(text) => Right(text.trim)
              case Left(_)     => Left("STT API returned unexpected format")
          case None => Left("Empty response from STT API")
    }.handleErrorWith { e =>
      logger.warn(s"STT transcription failed: ${e.getMessage}").as(Left(e.getMessage))
    }
  end transcribeViaChat

  private def transcribeViaMultipart(
      wavBytes: Array[Byte],
      language: Option[String]
  ): IO[Either[String, String]] =
    // Build multipart/form-data body
    val boundary = "----nebflow-stt-" + System.currentTimeMillis().toHexString
    val header = s"--$boundary\r\n".getBytes(StandardCharsets.UTF_8)
    val footer = s"\r\n--$boundary--\r\n".getBytes(StandardCharsets.UTF_8)

    // file part
    val filePartHeader =
      s"Content-Disposition: form-data; name=\"file\"; filename=\"audio.wav\"\r\n" +
        "Content-Type: audio/wav\r\n\r\n"
    val filePartHeaderBytes = filePartHeader.getBytes(StandardCharsets.UTF_8)

    // model part
    val modelPartHeader =
      s"--$boundary\r\n" +
        "Content-Disposition: form-data; name=\"model\"\r\n\r\n" +
        s"$model\r\n"
    val modelPartBytes = modelPartHeader.getBytes(StandardCharsets.UTF_8)

    // language part (optional)
    val langBytes = language match
      case Some(lang) =>
        (s"--$boundary\r\n" +
          "Content-Disposition: form-data; name=\"language\"\r\n\r\n" +
          s"$lang\r\n").getBytes(StandardCharsets.UTF_8)
      case None => Array.emptyByteArray

    // Combine all parts
    val bodyBytes: Array[Byte] =
      header ++ modelPartBytes ++ langBytes ++ filePartHeaderBytes ++ wavBytes ++ footer

    val request = HttpRequest
      .newBuilder()
      .uri(URI.create(endpoint))
      .header("Authorization", s"Bearer $apiKey")
      .header("Content-Type", s"multipart/form-data; boundary=$boundary")
      .POST(HttpRequest.BodyPublishers.ofByteArray(bodyBytes))
      .timeout(java.time.Duration.ofSeconds(30))
      .build()

    IO.blocking {
      val response = client.send(request, HttpResponse.BodyHandlers.ofString())
      if response.statusCode() != 200 then
        // 语句位在 IO.blocking 内——必须 *Sync（原裸 logger.warn 是死日志，本批顺手复活）
        logger.warnSync(s"STT API error ${response.statusCode()}: ${response.body().take(300)}")
        Left(s"STT API error: ${response.statusCode()}")
      else
        parse(response.body()).toOption match
          case Some(json) =>
            json.hcursor.downField("text").as[String] match
              case Right(text) => Right(text.trim)
              case Left(_) => Left("STT API returned unexpected format")
          case None =>
            // Some APIs return plain text
            val text = response.body().trim
            if text.nonEmpty then Right(text) else Left("Empty response from STT API")
    }.handleErrorWith { e =>
      logger.warn(s"STT transcription failed: ${e.getMessage}").as(Left(e.getMessage))
    }
  end transcribeViaMultipart
end SttService

File: nebflow-rs/nebflow-gateway/src/stt_service.rs
//! STT (speech-to-text) service — mirrors Scala SttService.scala.
//!
//! OpenAI-compatible audio transcription API client.
//!
//! Config file: `~/.nebflow/stt-config.json`
//! ```json
//! {
//!   "apiKey": "your-key",
//!   "model": "glm-asr-2512",
//!   "endpoint": "https://open.bigmodel.cn/api/paas/v4/audio/transcriptions"
//! }
//! ```

use std::path::Path;
use std::time::Duration;

use tracing::{info, warn};

const DEFAULT_MODEL: &str = "glm-asr-2512";
const DEFAULT_ENDPOINT: &str = "https://open.bigmodel.cn/api/paas/v4/audio/transcriptions";

/// STT speech recognition service.
pub struct SttService {
    api_key: String,
    model: String,
    endpoint: String,
    client: reqwest::Client,
}

impl SttService {
    /// Load from the default config path (`~/.nebflow/stt-config.json`).
    /// Returns None when the file is missing or has no valid `apiKey`.
    pub fn create() -> Option<Self> {
        Self::create_from(&nebflow_core::config::data_root().join("stt-config.json"))
    }

    /// Load from an explicit config path. Separated from `create()` for
    /// testability (avoids depending on the real `$HOME`).
    pub fn create_from(path: &Path) -> Option<Self> {
        if !path.exists() {
            return None;
        }
        let raw = std::fs::read_to_string(path).ok()?;
        let svc = Self::parse_config(&raw);
        if svc.is_some() {
            info!("STT service initialized");
        }
        svc
    }

    /// Parse config JSON. Pure function for testability.
    /// Requires a non-empty `apiKey`; `model`/`endpoint` have defaults.
    pub fn parse_config(raw: &str) -> Option<Self> {
        let json: serde_json::Value = serde_json::from_str(raw).ok()?;
        let api_key = json
            .get("apiKey")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())?
            .to_string();
        let model = json
            .get("model")
            .and_then(|v| v.as_str())
            .unwrap_or(DEFAULT_MODEL)
            .to_string();
        let endpoint = json
            .get("endpoint")
            .and_then(|v| v.as_str())
            .unwrap_or(DEFAULT_ENDPOINT)
            .to_string();
        Some(Self::new(api_key, model, endpoint))
    }

    /// Construct directly (used by parse_config and tests).
    pub fn new(api_key: String, model: String, endpoint: String) -> Self {
        let client = reqwest::Client::builder()
            .connect_timeout(Duration::from_secs(10))
            .timeout(Duration::from_secs(30))
            .build()
            .unwrap_or_else(|_| reqwest::Client::new());
        Self {
            api_key,
            model,
            endpoint,
            client,
        }
    }

    /// Transcribe WAV audio bytes, returning the recognized text.
    /// `language` is an optional language code (zh, en, ...).
    pub async fn transcribe(
        &self,
        wav_bytes: &[u8],
        language: Option<&str>,
    ) -> Result<String, String> {
        if wav_bytes.is_empty() {
            return Err("Empty audio data".into());
        }

        let file_part = reqwest::multipart::Part::bytes(wav_bytes.to_vec())
            .file_name("audio.wav")
            .mime_str("audio/wav")
            .map_err(|e| e.to_string())?;
        let mut form = reqwest::multipart::Form::new()
            .text("model", self.model.clone())
            .part("file", file_part);
        if let Some(lang) = language.filter(|s| !s.is_empty()) {
            form = form.text("language", lang.to_string());
        }

        let response = self
            .client
            .post(&self.endpoint)
            .bearer_auth(&self.api_key)
            .multipart(form)
            .send()
            .await
            .map_err(|e| {
                warn!("STT transcription failed: {e}");
                e.to_string()
            })?;

        let status = response.status();
        let body = response.text().await.map_err(|e| e.to_string())?;
        if !status.is_success() {
            warn!("STT API error {status}: {}", &body[..body.len().min(300)]);
            return Err(format!("STT API error: {status}"));
        }

        // Prefer JSON `{"text": "..."}`; some providers return plain text.
        if let Ok(json) = serde_json::from_str::<serde_json::Value>(&body) {
            if let Some(text) = json.get("text").and_then(|v| v.as_str()) {
                return Ok(text.trim().to_string());
            }
            return Err("STT API returned unexpected format".into());
        }
        let text = body.trim();
        if text.is_empty() {
            Err("Empty response from STT API".into())
        } else {
            Ok(text.to_string())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn create_from_missing_file_returns_none() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("stt-config.json");
        assert!(SttService::create_from(&path).is_none());
    }

    #[test]
    fn parse_config_requires_api_key() {
        assert!(SttService::parse_config("{}").is_none());
        assert!(SttService::parse_config(r#"{"apiKey": ""}"#).is_none());
        assert!(SttService::parse_config("not json").is_none());
    }

    #[test]
    fn parse_config_applies_defaults() {
        let svc = SttService::parse_config(r#"{"apiKey": "k"}"#).unwrap();
        assert_eq!(svc.model, DEFAULT_MODEL);
        assert_eq!(svc.endpoint, DEFAULT_ENDPOINT);
    }

    #[test]
    fn parse_config_reads_custom_values() {
        let svc =
            SttService::parse_config(r#"{"apiKey": "k", "model": "m1", "endpoint": "http://x/y"}"#)
                .unwrap();
        assert_eq!(svc.model, "m1");
        assert_eq!(svc.endpoint, "http://x/y");
    }

    #[test]
    fn create_from_valid_file_returns_some() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("stt-config.json");
        std::fs::write(&path, r#"{"apiKey": "k", "model": "m1"}"#).unwrap();
        let svc = SttService::create_from(&path).unwrap();
        assert_eq!(svc.model, "m1");
    }

    #[tokio::test]
    async fn transcribe_rejects_empty_audio() {
        let svc = SttService::new("k".into(), "m".into(), "http://127.0.0.1:1/".into());
        let err = svc.transcribe(&[], None).await.unwrap_err();
        assert_eq!(err, "Empty audio data");
    }
}
